### PR TITLE
Fix Indirect Data Analysis not load multiple spectra from a single ws in F(Q) fit

### LINKED ISSUE #28527: Fix data loading in FQ fit
Currently the way data is loaded in FQ fit is inconsistent and confusing.

When the data is loaded it is converted from FWHM to HWHM behind the scenes. It is also not possible to load more than one spectra from a single workspace in multiple input mode. 

This leads to a secondary issue where the names for all output workspaces have HWHM maxima appended as this is the workspace actually fitted but this is confusing when an EISF fit is done.

### PR DESCRIPTION
**Pull request by Harrietbrown**
**Description of work**

This PR makes it so that the multiple workspaces tab on the indirect data analysis f(q) fit tab can accept multiple spectra from a single workspace. Previously if a second spectra was added from the same workspace it would not be accepted.

re: #28527

**To test:**

1. Load test data.
1. open Indirect Data Analysis
2. go to the F(Q) fit tab and switch to multiple input mode.
4. click add workspace, change from file to workspace.
5. add width data from workspace, add a different spectra from the same workspace.
6. change to a different workspace and add a spectra.
7. these spectra should all be availalbe to fit.
8. Try messing with it in other ways to get it to break.

Fixes #28527

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by SpencerHowells** [comment]
@Harrietbrown The interface should only read the 1 specified spectrum from a workspace. For example, If FWHM is specified it should only be able to read the spectrum with FWHM in the name. For multiple input it would be another workspace again with FWHM.

**Comment by Harrietbrown** [comment]
@SpencerHowells What about in cases with multiple FWHM in a workspace.

**Comment by SpencerHowells** [comment]
@Harrietbrown That is a new situation - I've not thought about it! If workspaces are from different instruments, say, it could have 1 or 2 FWHM or rarely 3.  It should not read 2 from the same workspace always 1. So it should be a choice of 1 or 2.

File: qt/scientific_interfaces/Indirect/JumpFitDataPresenter.cpp [hunk 1]
```diff
@@ -202,6 +206,19 @@ void JumpFitDataPresenter::addWorkspace(IndirectFittingModel *model,
 void JumpFitDataPresenter::addDataToModel(IAddWorkspaceDialog const *dialog) {
   if (const auto jumpDialog =
           dynamic_cast<JumpFitAddWorkspaceDialog const *>(dialog)) {
+    //  update active data index with correct index based on the workspace name
+    //  and the vector in m_fitDataModel which is in the base class
+    //  indirectFittingModel get table workspace index
+    const auto wsName = jumpDialog->workspaceName().append("_HWHM");
```
Comment (Mantid-Matthew) [inline qt/scientific_interfaces/Indirect/JumpFitDataPresenter.cpp:212]: Not to do with this PR but for multiple selection we could solve the EISF having the wrong name by only doing this appending step in the HWHM case. Propbably better ways but though  would note it here as it occured.

**Comment by SpencerHowells** [comment]
@Harrietbrown Here are a couple of files for testing. 
[FqFit_Result.zip](https://github.com/mantidproject/mantid/files/5061216/FqFit_Result.zip)
The Conv file has just 1 FWHM/EISF. The QLR file has 6 spectra - choose f1.f1.FWHM/EISF.

File: qt/scientific_interfaces/Indirect/JumpFitModel.cpp [hunk 2]
```diff
@@ -243,8 +243,7 @@ void JumpFitModel::addWorkspace(const std::string &workspaceName) {
 
   const auto hwhmWorkspace =
       createHWHMWorkspace(workspace, name, parameters.widthSpectra);
-  IndirectFittingModel::addWorkspace(hwhmWorkspace->getName(),
-                                     Spectra(createSpectra(spectrum.get())));
+  IndirectFittingModel::addWorkspace(hwhmWorkspace->getName(), Spectra(""));
```
Comment (Mantid-Matthew) [inline qt/scientific_interfaces/Indirect/JumpFitModel.cpp:246]: What is the reasoning behind this change? I think adding a worspace to the fitting data within any Spectra specified may be at least partially responsible for one of the issues below.

**Review by Matthew-Andrew — CHANGES_REQUESTED** [review]
Code changes look good and I can now load multiple FWHM or EISF spectra from the same workspace if I wish, along with spectra from other workspaces and subsequently perform fits on them.

 I found several action sequences which led to the GUI being in odd states or crashing however.  I suspect several of these were in existence before this PR as multiple input mode has never quite worked with the F(Q) tab, and is very stateful and brittle. I list them here in case they are new or can be easily fixed at the same time as this PR.

 * If I attempt to add spectra from a workspace which contains both FWHM and EISF data. I I loud a FWHM maxima spectra and then switch to EISF the list is blank (this is probably ok as we don't want to load both types at once). If I then switch back to Width however the parameter name list remains blank.
 * The remove button does not curretly appear to do anything at least visually. I ran into some weird behavior after pressing it so is possible that it changes the hidden state of the GUI though.
 * If I load a workspace in single input mode and then switch to multiple and select this workspace rather than loading it the parameter type and parameter name combo boxes remain blank at least for workspaces with multiple spectra.
 * The following sequence leads to an error and hard crash:
    * Load worspace 1 add a spectra.
    * Load workspace 2 add a spectra
    * Switch back to workspace 1 using the workspace combo box option and load another spectra (can be the same). 
    * Above the plot click the combo box to select which workspace you are interested in. There are multiple entries for workspace 1. One of which has no spectra select this.
  * When first loading data from INI16B_125878_QLd_Results.nxs I repeatedly ran into a warning box saying Spectrum index 10 not found` but I could not reproduce this after a mantid restart so not ure what is going on here.

File: qt/scientific_interfaces/Indirect/JumpFitModel.cpp (hunk 2 above)
Comment (Harrietbrown) [reply qt/scientific_interfaces/Indirect/JumpFitModel.cpp:246]: When the workspace is added initially it adds all the spectra to the spectra list in the model. Previously this was replaced when a specra was added, but with the method now appending spectra to the list without this change it loads all the sectra as well as a double of the one you wanted.

**Comment by Harrietbrown** [comment]
> * If I attempt to add spectra from a workspace which contains both FWHM and EISF data. I I loud a FWHM maxima spectra and then switch to EISF the list is blank (this is probably ok as we don't want to load both types at once). If I then switch back to Width however the parameter name list remains blank.

This was already in the nightly

> * The remove button does not curretly appear to do anything at least visually. I ran into some weird behavior after pressing it so is possible that it changes the hidden state of the GUI though.

This was already in the nightly

> * If I load a workspace in single input mode and then switch to multiple and select this workspace rather than loading it the parameter type and parameter name combo boxes remain blank at least for workspaces with multiple spectra.

I can't seem to replicate this. If I load a workspace with multiple fwhm then switch to multiple and add workspace, the combo boxes fill as they should.

> * The following sequence leads to an error and hard crash:
>   
>   * Load worspace 1 add a spectra.
>   * Load workspace 2 add a spectra
>   * Switch back to workspace 1 using the workspace combo box option and load another spectra (can be the same).
>   * Above the plot click the combo box to select which workspace you are interested in. There are multiple entries for workspace 1. One of which has no spectra select this.

I'll try and sort this.

> * When first loading data from INI16B_125878_QLd_Results.nxs I repeatedly ran into a warning box saying Spectrum index 10 not found` but I could not reproduce this after a mantid restart so not ure what is going on here.

File: qt/scientific_interfaces/Indirect/JumpFitModel.cpp [hunk 3]
```diff
@@ -281,28 +280,47 @@ std::string JumpFitModel::getFitParameterName(TableDatasetIndex dataIndex,
...
-        dataIndex);
+    if (single == true) {
+      setSpectra(
+          createSpectra(std::vector<std::size_t>({widthSpectra[widthIndex]})),
+          dataIndex);
+    } else { // In multiple mode the spectra needs to be appending on the
+             // existing spectra list.
+      auto spectra_vec = std::vector<std::size_t>({widthSpectra[widthIndex]});
+      auto spectra = getSpectra(dataIndex);
+      for (size_t i = 0; i < spectra.size().value; i++) {
```
Comment (Mantid-Matthew) [inline qt/scientific_interfaces/Indirect/JumpFitModel.cpp:296]: Does the spectra object allow for each loops? My memory is that it doesn't buy if it does could use one here. I think the Spectra object going forward is one which could potentially be simplified. There are some edge cases to handle but a vector of int's would work for most cases.

File: qt/scientific_interfaces/Indirect/JumpFitModel.cpp [hunk 4]
```diff
@@ -281,28 +280,47 @@ std::string JumpFitModel::getFitParameterName(TableDatasetIndex dataIndex,
...
+      }
+      setSpectra(createSpectra(spectra_vec), dataIndex);
+    }
   } else
     logger.warning("Invalid width index specified.");
 }
 
 void JumpFitModel::setActiveEISF(std::size_t eisfIndex,
-                                 TableDatasetIndex dataIndex) {
+                                 TableDatasetIndex dataIndex, bool single) {
```
Comment (Mantid-Matthew) [inline qt/scientific_interfaces/Indirect/JumpFitModel.cpp:306]: How easy would it be to add some unit tests of this functionality?

File: qt/scientific_interfaces/Indirect/JumpFitDataPresenter.cpp [hunk 5]
```diff
@@ -224,9 +233,26 @@ void JumpFitDataPresenter::setModelSpectrum(int index) {
...
+    m_jumpModel->setActiveWidth(static_cast<std::size_t>(index), m_dataIndex,
+                                false);
   else
-    m_jumpModel->setActiveEISF(static_cast<std::size_t>(index), m_dataIndex);
+    m_jumpModel->setActiveEISF(static_cast<std::size_t>(index), m_dataIndex,
+                               false);
+}
+
+void JumpFitDataPresenter::setDataIndexToCurrentWorkspace(
+    IAddWorkspaceDialog const *dialog) {
```
Comment (Mantid-Matthew) [inline qt/scientific_interfaces/Indirect/JumpFitDataPresenter.cpp:244]: How easy would it be to add some unit tests of this functionality?

**Review by Matthew-Andrew — COMMENTED** [review]
This now works nicely for me good job. I've put a few code comments/questions below.

I also spotted two minor oddities which I don't think it's worth fixing here unless doing so is very easy. As this PR is already a large improvement on the current state of things.
 * It's possible to add the same spectra from a workspace multiple times. This briefly appears in the table before disapearing when you switch which workspace you are plotting. Is it possible to check and not add the same workspace index more than once for each workspace?
 * I see some hard to track down oddities when I add both width and EISF data. In my case this led to numerous different things with one workspaces data disapearing from the table entirely and the inability to add more data even after I closed and re-opened the GUI. A user will probably never do this but may simplify things if we don't allow adding both types of spectra at once.

**Comment by SpencerHowells** [comment]
@Harrietbrown @Mantid-Matthew It should never add Width & EISF. They are 2 different functions.

**Comment by Harrietbrown** [comment]
The tests are not a simple thing to add here. When the mainanance refactoring is finished the unit tests will be eaisier to make changes too and I can write for them then.

As for the issue with multiple spectra being gotten rid of when changing the workspace in the dialogue. I think we can leave that for a sepeate issue instead of add something else before the end of the sprint.

**Review by Matthew-Andrew — APPROVED** [review]
God job I think this is now good to go. This is a significant improvement in the functionality of the F(Q) multiple input options. There are still some issue surrounding this functionality but they all pre-date this PR.

@Harrietbrown could we make issues for the problems not fixed in this PR.

**Comment by gemmaguest** [comment]
@Harrietbrown Could you confirm whether you've created issues for the problems not fixed in this PR? Is it also worth creating an issue for the missing tests?

**Comment by Harrietbrown** [comment]
Issue #29244 made for the duplicate spectra issue

**Pull request merged**